cni-plugin/linkerd-cni-validator/src/main.rs

```diff
@@ -0,0 +1,169 @@
+            Ok(())
+        }
+
+        _ = shutdown_tx.signaled() => {
+            if let Err(e) = server_task.await.expect("Failed to run outbound server task") {
+                error!(error = %e, "Failed to validate outbound routing configuration");
+                std::process::exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+            error!("Failed to validate due to server terminating early");
+            std::process::exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready_rx: tokio::sync::oneshot::Receiver<Result<()>>,
+) -> Result<()> {
+    let timeout = std::time::Duration::from_secs(120);
+    if (tokio::time::timeout(timeout, ready_rx).await).is_err() {
+        anyhow::bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!(original_dst = %target_addr, "Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        let peer = socket.peer_addr()?;
+        debug!(original_dst = %peer, "Client connected to validation server");
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            if let Err(e) = is_readable {
+                anyhow::bail!("cannot read off client socket {}", e);
+            }
+        }
+
+        () = tokio::time::sleep(timeout) => {
+            anyhow::bail!("timed-out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
+
+    let mut buf = [0u8; 100];
+    let read_sz = stream.read(&mut buf[..REDIRECT_RESPONSE.len()]).await?;
+    let resp = String::from_utf8(buf[..REDIRECT_RESPONSE.len()].to_vec())?;
+    debug!(redirect_response = %resp, bytes_read = %read_sz);
+
+    if resp == REDIRECT_RESPONSE {
+        Ok(())
+    } else {
+        anyhow::bail!(
+            "expected client to receive {:?}, got {:?} instead",
+            REDIRECT_RESPONSE,
+            resp,
+        );
+    }
+}
+#[tracing::instrument(name = "outbound_server", skip(ready_tx, shutdown))]
+async fn outbound_serve(
+    listen_addr: SocketAddr,
+    ready_tx: tokio::sync::oneshot::Sender<Result<()>>,
+    shutdown: kubert::shutdown::Watch,
+) -> Result<()> {
+    let listener = TcpListener::bind(listen_addr)
+        .await
+        .expect("Failed to bind server");
+    info!(%listen_addr, "Listening for incoming connections");
+
+    if ready_tx.send(Ok(())).is_err() {
+        error!("Failed to send 'ready' signal, receiver dropped");
+        anyhow::bail!("Failed to bind server");
+    }
+
+    let resp_bytes = REDIRECT_RESPONSE.as_bytes();
+    tokio::spawn(handle_conn(listener, resp_bytes));
+
+    let _handle = shutdown.signaled().await;
+    debug!("Received shutdown signal");
+    Ok(())
+}
+
+#[tracing::instrument(name = "outbound_handler", skip_all)]
+async fn handle_conn(listener: TcpListener, resp_bytes: &'static [u8]) {
```

Comment:
naming nit: to me, `handle_conn` feels like what i would have called the function that handles a _single_ connection, while this function is actually the entire accept loop...

Reply:
good point. I re-named it to `accept` which should be pretty self-explanatory here but I'm more than happy to change if you have any suggestions.